src/python/pants/engine/target.py

```diff
@@ -48,39 +49,62 @@
 # -----------------------------------------------------------------------------------------------
 
 # Type alias to express the intent that the type should be immutable and hashable. There's nothing
-# to actually enforce this, outside of convention. Maybe we could develop a MyPy plugin?
-ImmutableValue = Any
+# to enforce immutability, outside of convention. Maybe we could develop a MyPy plugin?
+ImmutableValue = Hashable
 
+_DefaultBase = TypeVar("_DefaultBase", bound=Optional[ImmutableValue])
 
-class Field(ABC):
+
+# This @dataclass declaration is necessary for the `value` field to be propagated to the __hash__ of
+# all subclasses, but MyPy doesn't recognize generic abstract dataclasses yet.
+@dataclass(unsafe_hash=True)  # type: ignore[misc]
+class Field(Generic[_DefaultBase], metaclass=ABCMeta):
+    value: _DefaultBase
     # Subclasses must define these.
     alias: ClassVar[str]
-    default: ClassVar[ImmutableValue]
+    default: ClassVar[_DefaultBase]
     # Subclasses may define these.
     required: ClassVar[bool] = False
 
     # This is a little weird to have an abstract __init__(). We do this to ensure that all
-    # subclasses have this exact type signature for their constructor.
-    #
-    # Normally, with dataclasses, each constructor parameter would instead be specified via a
-    # dataclass field declaration. But, we don't want to declare either `address` or `raw_value` as
-    # attributes because we make no assumptions whether the subclasses actually store those values
-    # on each instance. All that we care about is a common constructor interface.
+    # subclasses have this exact type signature for their constructor. Normally we would rely on the
+    # generated __init__ by declaring `address` as a field, but we don't want that on *every* Field
+    # subclass.
     @abstractmethod
     def __init__(self, raw_value: Optional[Any], *, address: Address) -> None:
         pass
```

Comment:
It's probably fine to have a default `__init__` now. Then have `AsyncField` do this:

```python
super().__init__(...)
self.address = addresss
```

I think this would simplify things.

--

One weird edge, though. We ideally keep the `__init__()` as `@final`, as we don't expect subclasses beyond `PrimitiveField` and `AsyncField` to override it. If `PrimitiveField` is no longer defining `__init__()`, then we can't annotate with `@final` there.

I think we can mark it here, and then simply use a type ignore in the `__init__()` of `AsyncField`.

Reply:
Speaking of this...I'm not sure `PrimitiveField` is pulling its weight anymore! Thoughts on removing it? Every field is a `Field`, and a couple are special `AsyncField`s.

I think this would remove a lot of complexity.

Reply:
I absolutely think so! I will remove `PrimitiveField` (which was a weirdly specific name for a type anyway) and use the method you described to handle the `@final` __init__().

Reply:
This required a small hack in `meta.py` to unfreeze the base class, but with the added comment I still think this change is a good one.